### ISSUE TITLE
Regular Jump Problems

### ISSUE THREAD

**Issue by ChrisRackauckas** [JuliaDiffEq/DifferentialEquations.jl#250]
Our jump problems were built around jump-adapted algorithms. Jump-adapted algorithms step directly to the time of the jump. Our interface of `rate(u,p,t)` and `affect!(integrator)` took advantage of this because, since it's only one jump at a time and it has the integrator, it can basically do anything. I built it like this to do Poisson-distributed jumps where at the jump points I would change the number of cells in a MultiScaleArray.

Super unique application, yay. But the issue is that it's not in a form which can be used with a "regular" algorithm. A regular algorithm (in Platen's terminology, so finance terms) is an algorithm which does not directly step to the jump points. In finance this is the Strong Euler algorithm of Platen, in biology this is the tau-leaping method, etc. all the same deal. These need to calculate the number of times the jump will be taken in an interval and then the total effect of the jumps summed together. The current `rate(u,p,t)` setup works fine for that so that can stay. However, the setup with mutation on the integrator doesn't necessarily work with it: what exactly did it change and when? It can do all sorts of weird things, and it's hard to calculate "the total change" that the user does with a bunch of `affect!` calls.

However, we can remedy this by defining a different set of jumps. A `RegularJump` is defined by a rate `rate(u,p,t)` and a change function `c(u,p,t,zeta)` where zeta is the corresponding mark of the Poisson measure. This would then need a mark function `mark(u,p,t) = nothing` by default. Out-of-place is `c(u,p,t,zeta) = du`, and in-place is `c(du,u,p,t,zeta)` (does the mark need an in-place form?). Basically, it computes how `u` changes with the jump. A `JumpProblem` then can contain these as well, and a `JumpProblem` with these kinds of jumps will error if the solver cannot handle regular jumps. Then there's an easy conversion to `ConstantRateJump` and `VariableRateJump` of course.

This keeps constant and variable rate jumps as callback-dependent jump-adapted methods, but gives us a constrained option to define a simpler form of jumps that allow for a larger set of methods. One other thing to note is that to solve for the Kolmogorov equations of Levy processes you need jumps of this form. The jumps from the reaction networks fit this form so that would then build these kinds of jumps, which would then allow things like tau-leaping algorithms on them. This also makes it natural to mix some jumps as jump-adapted and some with a regular discretization.

Looking for advice here to see if this all makes sense. @gabrielgellner @onoderat @korsbo @gaussia @alanderos91 @elevien @jlperla

**Comment by gabrielgellner** [comment]
I think an inplace form for the mark function could make sense. If you have a lot of rates but only some subset updates on a step it could be useful to not have to recompute everything (unless I am missing exactly what this would be used for).

**Comment by onoderat** [comment]
I do not really know the math behind these jump process so I don't think I am qualified to comment on the implementation. 

I can point out that such a method will be very useful for the quantum optics community because jump process are behind describing quantum system under photon counting. When the number photons hitting the detector becomes high, the community traditionally just call it quits because the dt has to be reduced significantly for the number of clicks in each interval not exceed 1. This often forces us to study quantum system that leaks less photons -> limiting the parameter space of study.

Reference: http://qutip.org/docs/4.1/guide/dynamics/dynamics-monte.html

**Comment by alanderos91** [comment]
I confess that the full scope of your idea escapes me. Please read on at your own risk!

My understanding is that `rate(u,p,t)` should be able to handle tau-leaping as-is. The difficulty is in getting `affect!` and the integrator to play nicely with, for example, a leap step and arbitrary callbacks/events? With this understanding, the mark function is presumably used to make sense of the jump points for `affect!` without stepping through each one individually. I'm not sure this approach can completely account for the complications inherent in arbitrary calls to `affect!`, as you stated. For example, tau-leaping in the context of reaction networks operates under the assumption that each reaction's intensity is approximately constant over a leap. I believe event handling is a delicate matter in tau-leaping for this very reason.

> I think an inplace form for the mark function could make sense. If you have a lot of rates but only some subset updates on a step it could be useful to not have to recompute everything (unless I am missing exactly what this would be used for).

I agree. I don't think it would matter too much for tau-leaping algorithms without events, but it may be helpful in dealing with the integrator.

**Comment by jlperla** [comment]
Sorry, most of this is beyond my understanding for many reasons - in part because the economics and macro-finance literature I work in does not use trajectories/simulations/etc.

> One other thing to note is that to solve for the Kolmogorov equations of Levy processes you need jumps of this form.

But I know a little bit more about Kolmogorov Forward Equation with a jump-diffusion process, and I don't understand how you can step ahead to the time of the jump for that if you start with an initial condition as a whole distribution rather than a point.  In any strictly positive time step, there is a positive measure getting a jump?  This might just come down to solving these equations through trajectories vs. discretizing and solving the PDEs/ODEs directly.

For the Kolmogorov Backwards equation, using some variation of Feynman–Kac_formula takes the expectation that economists are interested in, and converts it to PDEs which are typically then discretized and solved.  Often there is an optimal control problem embedded in it.  Again, I don't see how you can know the time of the step or even have trajectories with optimal control around then, but maybe it would be a great approach to take in the future.

For what it is worth, the majority of applications in economics and macro-finance effectively use upwind finite-differences to convert the stochastic process into a markov chain and then solve optimal control problems using the markov chain .  A classic reference is https://www.amazon.ca/dp/0387951393/ref=pe_386430_30332290_TE_3p_dp_1

**Comment by ChrisRackauckas** [comment]
>But I know a little bit more about Kolmogorov Forward Equation with a jump-diffusion process, and I don't understand how you can step ahead to the time of the jump for that if you start with an initial condition as a whole distribution rather than a point. In any strictly positive time step, there is a positive measure getting a jump? This might just come down to solving these equations through trajectories vs. discretizing and solving the PDEs/ODEs directly.

That's very different. For one, I want to address non-constant initial conditions in a separate issue ( https://github.com/JuliaDiffEq/DifferentialEquations.jl/issues/251 ). An "initial condition" is just a Dirichlet distribution initial condition in the PDE. So assuming we have that, yes, `ConstantRateJump`s and `VariableRateJump`s are incompatible with the PDE formulation because they require jump-adapted stepping which is only well-defined for trajectory simulations (strong simulations). They need to be at the correct time point and apply their `affect!` to the integrator. Weak simulations (simulations of probability distributions) instead have probabilities that jumps occurred in given times. This is compatible with a `rate(u,p,t)` that causes discrete changes of `c(u,p,t,zeta) = du`. Thus these new `RegularJump`s are what you want for PDEs (and they match the definition of jump diffusions in mathematical finance literature like Platen's). For the conversion of a jump diffusion to the associated PDE I think it would be best to error if it sees any `ConstantRateJump` or `VariableRateJump`s.

**Comment by ChrisRackauckas** [comment]
@gabrielgellner @alanderos91 I'm curious: what do you usually use a mark for? I've never had a model with them. If it's just counting the number of jumps, then we could pass the integer `i` for how many times the jump has been called into the user's `c` function and they can use that in their `c` function to do jump-specific differences. Would that make sense?

**Comment by jlperla** [comment]
> Thus these new RegularJumps are what you want for PDEs (and they match the definition of jump diffusions in mathematical finance literature like Platen's). For the conversion of a jump diffusion to the associated PDE I think it would be best to error if it sees any ConstantRateJump or VariableRateJumps

Yeah, some of the math finance literature is slightly different from the economics and macro-finance literature in that they often don't embed control problems in their setups.  When you do that, simulating trajectories becomes a viable approach (as far as I can tell).

> This is compatible with a rate(u,p,t) that causes discrete changes of c(u,p,t,zeta) = du. Thus these new RegularJumps are what you want for PDEs (and they match the definition of jump diffusions in mathematical finance literature like Platen's). For the conversion of a jump diffusion to the associated PDE I think it would be best to error if it sees any ConstantRateJump or VariableRateJumps.

It is pretty important in economics and finance to allow both a state/time varying arrival rates and a state/time varying jump size conditional on the jumps.  I think this is what you are referring to above, but just wanted to check.

What kinds of boundary values are you planning for the jump-diffusions?  This is where I start to get confused, but I know that they are covered in the Kushner and Dupuis book I mentioned, and in the following book (http://homepages.math.uic.edu/~hanson/pub/Slides/bk07fkebkefinal.pdf Section 7.6)

**Comment by ChrisRackauckas** [comment]
>It is pretty important in economics and finance to allow both a state/time varying arrival rates and a state/time varying jump size conditional on the jumps. I think this is what you are referring to above, but just wanted to check.

Yes, they are state/parameter/time varying arrival rates (`rate(u,p,t)`) and jump sizes (that's the `c(u,p,t,zeta)`).

>What kinds of boundary values are you planning for the jump-diffusions?

The boundary conditions only apply to PDEs. For trajectory simulation the domain of the state variables doesn't have to be specified.

That link uses a notation which is compatible with DiffEq's, so let me explain this a little bit. We have

```julia
sde_prob = SDEProblem(f,g,u0,tspan)
```

which defines the SDE. From the SDE you can do a few things. You can directly `solve` that to get trajectories. You can define jumps and then merge the jumps with the SDE to form a `JumpProblem` (i.e. a JumpDiffusion). If you only use `RegularJump(dp,h)` (here, `dp = rate` and `h = c`) jumps (and you can declare finitely many), then this is of the form in that paper (if you use other jumps, it's in the extended form). This can be directly simulated for trajectories via `solve`.

From there, on either the `SDEProblem` or the `JumpProblem` types, we will define a `KolmogorovBackward(prob,...)` and `KolmogorovForward(prob,...)` that spit out `DiffusionAdvectionProblem` types, i.e. do the conversion of the SDE/jump equation to the PDE. The `...` are all of the necessary extra things to define the PDE, which includes the domain and the boundary conditions. Some others like `StationaryDistribution(prob,...)` can be defined to which spit out the appropriate PDE problem types.

That problem then lives in DiffEq's PDE area as one of the many `PDEProblem` types we are going to offer. Solvers for that PDE can then be applied to the problem. They don't necessarily even need to know that it came from an SDE or a jump diffusion, they just solve the resulting PDE. These solvers are what will be built using things like DiffEqOperators and ApproxFun along with the ODE solvers.

**Comment by alanderos91** [comment]
I've never worked with the marks in tau-leaping, at least not explicitly. For my simulator, each reaction channel's intensity, along with a leap increment, is used as the rate parameter in a Poisson distribution. For `k` reactions, the number of times each channel fires over [0,tau] is `Y[j] = rand(Poisson(a[j] * tau))`. The marks are just the labels `j=1:k` used to denote which reaction fired. The ordering of the firing events is immaterial in the well-mixed case so the marks are irrelevant.

Knowing the order and number of jumps is probably all you need. You can probably impute an ordering without computing any inter-arrival times.

**Comment by ChrisRackauckas** [comment]
It looks like from @jlperla 's link and Platen's work that the mark is a random variable, so it should be `mark_distribution` and then we just sample from that (or default to nothing). Note that the setup allows multiple jump processes, so this isn't like choosing which reaction happened. 

I am not sure that order of the jumps is needed for this at all, just the number of each jump which is Poisson distributed. Where in the algorithm is the ordering necessary?

**Comment by alanderos91** [comment]
For tau-leaping: given you have N_t total reaction firings over [0,t], one can determine the average number of firings for each reaction based on the finite-time transition probabilities p_j = a_j(x) / a_0(x) using a marked Poisson process. This works nicely because the random variable for the marking process is just a categorical random variable.

The markings and ordering are *not* required for tau-leaping *without events*. As you stated, you only need to know the number calls for each jump because the updates can be lumped together and done in parallel.

However, this perspective may be relevant for event handling if the events depend on time (I do not fully understand how event handling works in DifferentialEquations.jl). A (contrived) example may be administering a drug at regular intervals. Would the event trigger at the end of a leap, or during the leap? The effect may be dramatic and violate the leap condition one imposes before the leap.

**Comment by ChrisRackauckas** [comment]
Do you have a source? I am not getting the intuition of the marks here.

>A (contrived) example may be administering a drug at regular intervals. Would the event trigger at the end of a leap, or during the leap? 

Well there's two options. If it's a DiscreteCallback, `dt` would be adjusted before the step, and the event would happen after the step so then tau-leaping doesn't need to know about it. If it's a ContinuousCallback, it will happen in the middle of the step and then interpolation will pull back to the event point. That doesn't seem to make sense for jump problems which don't use continuous variables, but I guess this must be what the marks are for?

**Comment by alanderos91** [comment]
I see, so one checks if any events trigger over the leap interval [t, t + dt] and adjusts dt accordingly. If "continuous" always refers to the state variables, then I agree that ContinousCallback never makes sense for jump problems and this is a non-issue altogether. I simply thought the marks could be helpful in handling events that require interpolation but I now suspect this is not the case.

Another example: If an event depends on some state variable `X` hitting a value `theta` and a leap of size `dt` results in `X >= theta`, when does the event trigger? There may be many paths from `X != theta` to `X == theta`. The tau-leaping procedure loses information about the path taken.

> Do you have a source? I am not getting the intuition of the marks here.

Chapter 6 in Applied Probability: [https://www.springer.com/us/book/9781441971647](https://www.springer.com/us/book/9781441971647)

Relevant sections are 6.2 and 6.8. I'm describing "coloring" of a Poisson process, which is the simpler case of the more general notion of "marking".

**Comment by ChrisRackauckas** [comment]
>I see, so one checks if any events trigger over the leap interval [t, t + dt] and adjusts dt accordingly. If "continuous" always refers to the state variables, then I agree that ContinousCallback never makes sense for jump problems and this is a non-issue altogether. I simply thought the marks could be helpful in handling events that require interpolation but I now suspect this is not the case.

Continuous refers to the independent variable. I am just not understanding how one interpolates to solve this problem:

>If an event depends on some state variable X hitting a value theta and a leap of size dt results in X >= theta, when does the event trigger?

How does marking handle that?

**Comment by jlperla** [comment]
> It looks like from @jlperla 's link and Platen's work that the mark is a random variable, so it should be mark_distribution and then we just sample from that (or default to nothing). 

Yes.  Or to be careful, it needs to be a random variable which can be conditioned on both the state and time.

**Comment by alanderos91** [comment]
My mistake, I don't mean interpolation. The idea I had in mind is:

1. Given the total number of firings, compare the expected number of firings (marking, using the finite-time transition probabilities) for each reaction to the realized numbers in the leap.
2. Using the information from (1) and the reaction network structure, use some recipe to figure out a reasonable guess for the time s in [t, t + dt] such that the event triggers.
3. Compute the change up to time s, including the triggered event.
4. Compute the change over [t + s, t + dt].
5. Complete the leap.

> How does marking handle that?

It doesn't. I was hasty; the expected number of firings for each reaction over some interval [t, t + dt] does not tell you much (if anything) about how they are "spaced out".

**Comment by ChrisRackauckas** [comment]
I'm sure there's a way to derive a Poisson bridging distribution and a two-stack representation to do this properly, but I don't know of a derivation of how to properly do all of that yet. It would be a good next step after my SDE adaptivity paper to do this kind of time pullback for jump diffusions. I'll keep this in mind but for now we'll just have to not have exact even handling for jump systems, which is fine. 

But okay, it looks like this design will be able to handle that so that's good to keep in mind. Also `mark(u,p,t) = nothing` returns a distribution (or nothing) answers @jlperla 's question. I think we're good then. @jlperla , marks are scalars?

One last thing to address is the representation. Should they be separate `(rate,c,mark)` tuples, or represented singularly? It made sense to do `ConstantRateJump` and `VariableRateJump` separately because they had to have `affect!` functions, but in this case I don't think so. Like how `dW` is a vector and so `g*dW` does multiple noise processes, `rate(out,u,p,t)` could be a vector function with the size being the number of jumps, and then `c(dc,u,p,t,mark)` could be a matrix where column `dc[:,i]` are the changes due to jump `i`. Then `dc*j` for `j` the vector of jumps per jump process would give the update. The paper @jlperla linked (http://homepages.math.uic.edu/~hanson/pub/Slides/bk07fkebkefinal.pdf) uses this representation. 

The difference between the two is that the first would be a long tuple of functions, while the latter is a single vector function. The long tuple of functions likely won't scale well to large numbers of jumps (>16) since large tuples carry a ton of type information (compile times probably wouldn't be good there either). This would effectively limit the number of jump processes you'd want to use with it. The other case, in order to scale you'd want `dc` to modify elements of a sparse matrix. This is what we do with SDEs right now.

**Comment by ChrisRackauckas** [comment]
There are ways to work with the sparse matrix that are fast. If you directly modify `A.nzval` knowing the structure (it goes column-wise), then there's no overhead. So in theory any `c` function on the sparse matrix can be written without overhead there. Then sparse matrix multiplication is actually faster after a certain point:

```julia
julia> A = sprand(100,100,0.1);

julia> A2 = full(A);

julia> c = rand(100); b = similar(c);

julia> using BenchmarkTools

julia> A
100×100 SparseMatrixCSC{Float64,Int64} with 983 stored entries:
  [15 ,   1]  =  0.798806
  [26 ,   1]  =  0.218812
  [38 ,   1]  =  0.122372
  [41 ,   1]  =  0.293602
  [55 ,   1]  =  0.273528
  [69 ,   1]  =  0.391787
  [73 ,   1]  =  0.380623
  ⋮
  [56 , 100]  =  0.867886
  [60 , 100]  =  0.733264
  [61 , 100]  =  0.0326307
  [64 , 100]  =  0.608664
  [65 , 100]  =  0.0999586
  [71 , 100]  =  0.766003
  [91 , 100]  =  0.66284
  [98 , 100]  =  0.593513

julia> c = rand(100); b = rand(100);

julia> @btime A_mul_B!(c,A,b)
  927.000 ns (0 allocations: 0 bytes)
100-element Array{Float64,1}:
 1.86665
 3.9889
 2.31723
 2.3641
 1.71298
 1.09748
 1.72212
 1.85886
 1.23162
 2.21755
 ⋮
 2.48642
 1.6954
 2.75795
 2.58893
 2.8498
 2.12207
 2.97069
 2.31545
 2.18362
 2.70729

julia> @btime A_mul_B!(c,A2,b)
  22.541 μs (0 allocations: 0 bytes)
100-element Array{Float64,1}:
 1.86665
 3.9889
 2.31723
 2.3641
 1.71298
 1.09748
 1.72212
 1.85886
 1.23162
 2.21755
 ⋮
 2.48642
 1.6954
 2.75795
 2.58893
 2.8498
 2.12207
 2.97069
 2.31545
 2.18362
 2.70729
```

The sparse format via tuples recursively get `dc` from each of the `c` and then add the `dc`s in-place, so it would be more similar to the dense multiplication here (lots of extra adds!). So this is good proof that the matrix formulation will actually scale better to large reaction networks than the tuples, if people know how to modify sparse matrices well. @Gaussia 's reaction network DSL could setup a dispatch to use `A.nzval` updates and it would be absolutely killer for large sparse reaction networks (@korsbo we'll probably need your help for that). 

So then the proposal is for `RegularJump` to be a single matrix function where `rate` is a vector of rates and `h` (let's call it `h` since then it's `f`, `g`, `h`) is a matrix function where matrix multiplication applies the `dh*dp` to a vector of jump numbers. 

Since this won't interact well with other jumps, at least for now, we can just call this `JumpDiffusionProblem` and write solvers for that. I'd just stick them in StochasticDiffEq.jl as the future adaptive algorithm would make use of the same engine with a different form of `NoiseProcess`. We can also make `RegularJumpProblem` for jump-only problems (biological networks) and write algorithms specifically on that. I am inclined to put those algorithms in OrdinaryDiffEq.jl as an extension to `FunctionMap` and its discrete handling. StochasticDiffEq.jl has extra stuff for NoiseProcesses that I would want to avoid in any case which doesn't use a continuous noise process.

Bikeshed the naming scheme and then I think we have a plan.

**Comment by isaacsas** [comment]
Regarding Poisson bridging distributions for event handling, perhaps some modification of Anderson's post-leap checking method for avoiding negative population values in tau-leaping would work? His method essentially looks if the population is negative after a leap, and if so rejects the leap and attempts a smaller leap. However, it keeps track of Poisson rv samples and uses (binomial) bridging distributions to avoid bias in determining the number of events over the smaller leap step. To my knowledge, it is the only method that guarantees non-negativity of the tau-leaping method in an unbiased manner, however, it is much more complicated to implement than other leaping methods (so not as commonly used).

[https://www.math.wisc.edu/~anderson/papers/AndPostleap.pdf](url)

**Comment by ChrisRackauckas** [comment]
Not sure why your url doesn't work, but this does. 

https://www.math.wisc.edu/~anderson/papers/AndPostleap.pdf

>To my knowledge, it is the only method that guarantees non-negativity of the tau-leaping method in an unbiased manner, however, it is much more complicated to implement than other leaping methods (so not as commonly used).

Actually for us it's super easy to do since our stochastic differential equation is built around http://chrisrackauckas.com/assets/Papers/ChrisRackauckas-AdaptiveSRK.pdf which utilizes `NoiseProcess` distributions with an interpolating setup via bridging. This has already been refactored so that colored noise can happen, and all that needs to be given are the distribution and the bridging distribution:

https://github.com/JuliaDiffEq/DiffEqNoiseProcess.jl/blob/master/src/wiener.jl#L13-L29

If we setup a `NoiseProcess` with a Poisson and a Binomial, then it will naturally do Anderson's algorithm. That makes the implementation details clear then: the pure jump problems should be built on top of StochasticDiffEq.jl. This would extend to Poisson Runge-Kutta methods quite well too.

**Comment by ChrisRackauckas** [comment]
The basic `RegularJump` implementation is complete. It's implemented as a third type of jump, and unlike the others only one is allowed in a `JumpSet` because it pools. `JumpProblem`s which have a `RegularJump` require compatible algorithms since they are not implemented over the event handling system. I implemented 2 such algorithms: `SimpleTauLeaping` and `RegularSSA`. The latter is a version of SSA implemented directly on a `JumpProblem` with only `RegularJump`s. This is mostly a proof of concept though. 

Basically, `ConstantRateJump`s will still be the workhorse of SSAs since they allow things like re-ordering due to the sparse representation, and this is required for some algorithms like Gibson-Bruck SSA. `RegularJump`s are for jumps that you want to do with regular stepping, but also can be used to make hybrid methods. While you cannot re-order in this data format, that doesn't seem to be an issue because you'd only SSA that low rate parts in a hybrid algorithm anyways. Hybrid methods would just multiply the regular part via a vector of bools and control stepping, so it's surprisingly easy and efficient to implement like this.

StochasticDiffEq.jl will be used to build some adaptive regular jumping methods like @isaacsas recommended, and then some simple pure-jump integrators will be in DiffEqJump.jl

Future reference: https://gitter.im/JuliaDiffEq/Lobby/archives/2018/03/13
